### PR TITLE
[IOTDB-5649] Refactor `ExpressionAnalyzer` using visitor pattern

### PR DESCRIPTION
**Pull request by liyuheng55555**
ExpressionAnalyzer is the class in charge of handling expressions. 

Each method in this class has different ability, such as searchSourceExpressions() finds out which sources the expression contains, removeWildcardInExpression() remove * from expression, and so on.

There were some redundant logic in ExpressionAnalyzer's methods. To make it easier to add new kinds of expressions, it's necessary to simplify. 

After analyze the similarity between these methods, I believe the visitor pattern is a good way to make it simple. The same logic can be contained in superclass-visitor, and the difference can be implemented by subclass-visitor's override.

Finally, I restructure 11 methods in ExpressionAnalyzer, meantime add 11 new visitor. Here is their relationship:
|old method|new visitor as replacement|
|:--:|:--:|
|searchSourceExpressions|CollectSourceExpressionsVisitor|
|searchAggregationExpressions|CollectAggregationExpressionsVisitor|
|removeWildcardInExpression|RemoveWildcardInExpressionVisitor|
|removeWildcardInFilter|RemoveWildcardInFilterVisitor|
|replaceRawPathWithGroupedPath|ReplaceRawPathWithGroupedPathVisitor|
|concatDeviceAndRemoveWildcard|ConcatDeviceAndRemoveWildcardVisitor|
|removeWildcardInFilterByDevice|RemoveWildcardInFilterByDeviceVisitor|
|removeAliasFromExpression|RemoveAliasFromExpressionVisitor|
|getMeasurementExpression|GetMeasurementExpressionVisitor|
|bindTypeForTimeSeriesOperand|BindTypeForTimeSeriesOperandVisitor|
|concatExpressionWithSuffixPaths|ConcatExpressionWithSuffixPathsVisitor|

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/ReconstructVisitor.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.iotdb.db.mpp.plan.analyze.ExpressionUtils.reconstructAllKindsOfExpression;
+
+/**
+ * Collect result from child, then reconstruct. For example, two child each give me 1 result, I
+ * should use them to reconstruct 1 new result to upper level.
+ */
+public class ReconstructVisitor<C> extends ExpressionAnalyzeVisitor<Expression, C> {
```
Comment (liuminghui233) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/ReconstructVisitor.java:37]: make `ReconstructVisitor` abstract

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/RemoveWildcardInFilterVisitor.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+      if (actualPaths.isEmpty()) {
+        return Collections.singletonList(new NullOperand());
+      }
+      noStarPaths.addAll(actualPaths);
+    }
+    return reconstructTimeSeriesOperands(noStarPaths);
+  }
+
+  @Override
+  public List<Expression> visitLeafOperand(LeafOperand leafOperand, Context context) {
```
Comment (liuminghui233) [inline server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/RemoveWildcardInFilterVisitor.java:117]: It is confusing to have `visitTimeSeriesOperand()` and `visitLeafOperand()` at the same time. This is handled better in `RemoveWildcardInFilterByDeviceVisitor`, I think.

**Review by liuminghui233 — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionAnalyzer.java [hunk 3]
```diff
@@ -318,6 +250,9 @@ public static List<Expression> concatExpressionWithSuffixPaths(
    */
   public static List<PartialPath> concatExpressionWithSuffixPaths(
       Expression expression, List<PartialPath> prefixPaths) {
+    //    return new ConcatExpressionWithSuffixPathsVisitor().process(
+    //            expression, new ConcatExpressionWithSuffixPathsVisitor.Context(prefixPaths,
+    // null));
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/ExpressionAnalyzer.java]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/CartesianProductVisitor.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import org.apache.iotdb.db.mpp.plan.expression.binary.BinaryExpression;
+import org.apache.iotdb.db.mpp.plan.expression.ternary.TernaryExpression;
+import org.apache.iotdb.db.mpp.plan.expression.unary.UnaryExpression;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.iotdb.db.mpp.plan.analyze.ExpressionUtils.cartesianProductAllKindsOfExpression;
+
+public class CartesianProductVisitor<C> extends ExpressionAnalyzeVisitor<List<Expression>, C> {
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/CartesianProductVisitor.java]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/CollectVisitor.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  @Override
+  public List<Expression> visitUnaryExpression(UnaryExpression unaryExpression, Void context) {
+    return collectFromChild(unaryExpression);
+  }
+
+  @Override
+  public List<Expression> visitLeafOperand(LeafOperand leafOperand, Void context) {
+    return super.visitLeafOperand(leafOperand, context);
+  }
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/CollectVisitor.java]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/CollectVisitor.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import org.apache.iotdb.db.mpp.plan.expression.unary.UnaryExpression;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Simply collects result from child-expression. For example, two child give me 3 and 4 results, I
+ * will return 3+4 = 7 results to upper level.
+ */
+public class CollectVisitor extends ExpressionAnalyzeVisitor<List<Expression>, Void> {
```
[the remark this answers is not in the document]
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/CollectVisitor.java]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/ReconstructVisitor.java (hunk 1 above)
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/ReconstructVisitor.java]: done.

File: server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/RemoveWildcardInFilterVisitor.java (hunk 2 above)
Comment (liyuheng55555) [reply server/src/main/java/org/apache/iotdb/db/mpp/plan/expression/visitor/RemoveWildcardInFilterVisitor.java]: done.

**Review by liuminghui233 — APPROVED** [review]
LGTM :-)

**Pull request merged**